src/v/raft/group_configuration.cc

```diff
@@ -198,58 +224,163 @@ void erase_id(std::vector<vnode>& v, model::node_id id) {
+      *this);
+    make_change_strategy()->discard_old_config();
+}
+
+void group_configuration::abort_configuration_change(model::revision_id rev) {
+    vassert(
+      type() != configuration_type::simple,
+      "can not abort configuration change if it is of simple type - {}",
+      *this);
+    make_change_strategy()->abort_configuration_change(rev);
+}
+
+void group_configuration::cancel_configuration_change(model::revision_id rev) {
+    vassert(
+      type() != configuration_type::simple,
+      "can not cancel configuration change if it is of simple type - {}",
+      *this);
+    make_change_strategy()->cancel_configuration_change(rev);
+}
+
+void group_configuration::promote_to_voter(vnode id) {
+    auto it = std::find(
+      _current.learners.cbegin(), _current.learners.cend(), id);
+    // do nothing
+    if (it == _current.learners.cend()) {
+        return;
+    }
+    // add to voters
+    _current.learners.erase(it);
+    _current.voters.push_back(id);
+}
+
+bool group_configuration::maybe_demote_removed_voters() {
+    vassert(
+      _old,
+      "can not demote removed voters as configuration is of simple type - {}",
+      *this);
+
+    // no voters are present, do nothing
+    if (_old->voters.empty()) {
+        return false;
+    }
+    // if voter was removed, make it a learner
+    auto it = std::stable_partition(
+      _old->voters.begin(), _old->voters.end(), [this](const vnode& v) {
+          return _current.contains(v);
+      });
+
+    // nothing to remove
+    if (std::distance(it, _old->voters.end()) == 0) {
+        return false;
+    }
+
+    std::move(it, _old->voters.end(), std::back_inserter(_old->learners));
+    _old->voters.erase(it, _old->voters.end());
+
+    return true;
+}
+
+void group_configuration::update(model::broker broker) {
+    auto it = std::find_if(
+      _brokers.begin(), _brokers.end(), [id = broker.id()](model::broker& b) {
+          return id == b.id();
+      });
+
+    if (it == _brokers.end()) {
+        throw std::invalid_argument(fmt::format(
+          "broker {} does not exists in configuration {}", broker.id(), *this));
+    }
+
+    *it = std::move(broker);
+}
+
+/**
+ * Update strategy for v3 configuration
+ */
+
+void configuration_change_strategy_v3::add(
+  std::vector<model::broker> brokers, model::revision_id rev) {
+    _cfg._revision = rev;
     for (auto& b : brokers) {
         auto it = std::find_if(
-          _brokers.cbegin(),
-          _brokers.cend(),
+          _cfg._brokers.cbegin(),
+          _cfg._brokers.cend(),
           [id = b.id()](const model::broker& n) { return id == n.id(); });
-        if (unlikely(it != _brokers.cend())) {
+        if (unlikely(it != _cfg._brokers.cend())) {
             throw std::invalid_argument(fmt::format(
               "broker {} already present in current configuration {}",
               b.id(),
-              *this));
+              _cfg));
```

Comment:
i'm imagining that these exceptions, vasserts, logging, etc... will benefit from self-identifying their strategy version?

Reply:
sorry for not making it clear in the commit related with introducing change strategy. The change strategy is introduced to change the update logic based on the version of configuration. We log configuration version so it should be enough to identify the strategy that is being used.